### PR TITLE
HWY-159: Introducing maximum round exponent

### PR DESCRIPTION
**Pull request by xcthulhu**
https://casperlabs.atlassian.net/browse/HWY-159

### CODE REVIEW

File: node/src/components/consensus/highway_core/state/tests.rs [hunk 1]
```diff
@@ -173,11 +174,16 @@ fn add_vote() -> Result<(), AddVoteError<TestContext>> {
         .err()
         .map(vote_err);
     assert_eq!(Some(VoteError::InconsistentPanorama(BOB)), opt_err);
-    // And you can't change the round exponent within a round.
+    // And you can't make the round exponent too small
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state/tests.rs:177]: This was actually testing something else – that you cannot use different round exponent for confirmation/witness vote, only for the proposal. Although it indeed used `VoteError::RoundLength` for that. Let's see what @afck has to say about it.

**Review by goral09 — APPROVED** [review]
Neat. Great job.

File: node/src/components/chainspec_loader/chainspec.rs [hunk 2]
```diff
@@ -123,6 +125,10 @@ impl HighwayConfig {
         {
             warn!("Era duration is less than minimum era height * round length!");
         }
+
+        if self.minimum_round_exponent >= self.maximum_round_exponent {
+            warn!("Minimum round exponent is greater than or equal to maximum round exponent.  Minimum round exponent: {min}, Maximum round exponent: {max}", min = self.minimum_round_exponent, max = self.maximum_round_exponent);
```
Comment (afck) [inline node/src/components/chainspec_loader/chainspec.rs:130]:
```rust
    "You can break string literals using a \
    backslash!"
```

Comment (afck) [reply node/src/components/chainspec_loader/chainspec.rs:130]: And actually, maybe we should just panic here if it's strictly greater?
On the other hand if it's equal, maybe that's fine: You might want to run a network with a fixed block rate.

File: node/src/components/consensus/highway_core/state.rs [hunk 3]
```diff
@@ -528,7 +533,7 @@ impl<C: Context> State<C> {
                 // greater of the two exponents, a round boundary must be between the votes.
                 let max_re = prev_vote.round_exp.max(wvote.round_exp);
                 if prev_vote.timestamp >> max_re == timestamp >> max_re {
-                    return Err(VoteError::RoundLength);
+                    return Err(VoteError::RoundExpLengthLessThanMinimum);
```
Comment (afck) [inline node/src/components/consensus/highway_core/state.rs:536]: I agree with @goral09's comment below: This error is that someone changed their round exponent in the middle of a round. Maybe `VoteError::RoundLengthChangedWithinRound`?

File: node/src/components/consensus/highway_core/state/tests.rs (hunk 1 above)
Comment (afck) [reply node/src/components/consensus/highway_core/state/tests.rs:177]: Yes: Bob's round length is 16, and he's changing it to 32 in the next line, at time 50.
However, his previous vote `b1` was at time `49`, and there is no `49 < t ≤ 50` that is divisible by 32.

We should make `b1` above more explicit:
```rust
let b1 = add_vote!(state, rng, BOB, 49, 4u8, None; N, b0, c0)?;
```

File: resources/local/chainspec.toml [hunk 4]
```diff
@@ -53,6 +53,10 @@ finality_threshold_percent = 10
 # Integer between 0 and 255. The power of two that is the number of milliseconds in the minimum round length, and
 # therefore the minimum delay between a block and its child. E.g. 14 means 2^14 milliseconds, i.e. about 16 seconds.
 minimum_round_exponent = 12
+# Integer between 0 and 255. Must be greater than `minimum_round_exponent`. The power of two that is the number of
```
Comment (afck) [inline resources/local/chainspec.toml:56]: Maybe "greater or equal"?
And "the power of two of that"?

File: node/src/components/consensus/protocols/highway/round_success_meter.rs [hunk 5]
```diff
@@ -178,3 +180,106 @@ impl<C: Context> RoundSuccessMeter<C> {
...
+                break;
+            };
+            round_success_meter.current_round_id += 1;
+        }
+        assert_eq!(round_success_meter.new_exponent(), TEST_ROUND_EXP - 1);
+    }
+
+    #[test]
+    fn new_exponent_can_not_speed_up_because_min_round_exp() {
+        // If there's been enough successful rounds and its an acceleration round, but we are
```
Comment (afck) [inline node/src/components/consensus/protocols/highway/round_success_meter.rs:264]:
```suggestion
        // If there's been enough successful rounds and it's an acceleration round, but we are
```

File: node/src/components/consensus/protocols/highway/round_success_meter.rs [hunk 6]
```diff
@@ -178,3 +180,106 @@ impl<C: Context> RoundSuccessMeter<C> {
...
+            );
+        // If there have been more rounds of failure than MAX_FAILED_ROUNDS, slow down -- but can't
+        // slow down because of ceiling
+        round_success_meter.rounds = vec![false; MAX_FAILED_ROUNDS + 1].into();
+        assert_eq!(round_success_meter.new_exponent(), TEST_MAX_ROUND_EXP);
+    }
+
+    #[test]
+    fn new_exponent_speed_up() {
+        // If there's been enough successful rounds and its an acceleration round, speed up
```
Comment (afck) [inline node/src/components/consensus/protocols/highway/round_success_meter.rs:241]:
```suggestion
        // If there's been enough successful rounds and it's an acceleration round, speed up
```

File: node/src/components/consensus/protocols/highway/round_success_meter.rs [hunk 7]
```diff
@@ -178,3 +180,106 @@ impl<C: Context> RoundSuccessMeter<C> {
...
+            let current_round_index =
+                round_success_meter.current_round_id >> round_success_meter.current_round_exp;
+            if current_round_index % ACCELERATION_PARAMETER == 0 {
+                break;
+            };
+            round_success_meter.current_round_id += 1;
+        }
+        assert_eq!(round_success_meter.new_exponent(), TEST_MIN_ROUND_EXP);
+    }
+}
```
Comment (afck) [inline node/src/components/consensus/protocols/highway/round_success_meter.rs:285]: Awesome! Thanks for adding all those tests! :+1:

**Review by afck — COMMENTED** [review]
Looks great! But let's add another `VoteError` variant before merging.

File: node/src/components/chainspec_loader/chainspec.rs (hunk 2 above)
Comment (xcthulhu) [reply node/src/components/chainspec_loader/chainspec.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/12d3251cad27747c248568412894a5fc61524459 .  I've changed this to be a strict equality as you have suggested.

**Review by fizyk20 — COMMENTED** [review]
Looking good except the part pointed out by Mateusz regarding changes to the exponent mid-round.
Also, thanks for adding tests for `RoundSuccessMeter`!

File: node/src/components/consensus/highway_core/state.rs [hunk 8]
```diff
@@ -62,6 +62,8 @@ pub(crate) enum VoteError {
     InstanceId,
     #[error("The signature is invalid.")]
     Signature,
+    #[error("The round length exponent has somehow changed within a round.")]
+    RoundExpLengthChangedWithinRound,
```
Comment (afck) [inline node/src/components/consensus/highway_core/state.rs:66]: Either `RoundExpChangedWithinRound` or `RoundLengthChangedWithinRound`.
The length of the round exponent would be the logarithm of the logarithm of the round length… :smile: 
(Or maybe you meant `RoundLengthExpChangedWithinRound`?)

**Review by afck — APPROVED** [review]
Looks good! Just one more typo.

File: node/src/components/consensus/highway_core/state.rs (hunk 3 above)
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/state.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/13a22ef3964dd9ad4f38e73682e5d6f46e56bd12

File: node/src/components/consensus/highway_core/state.rs (hunk 8 above)
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/state.rs]: I meant `RoundLengthExpChangedWithinRound`.

Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/13a22ef3964dd9ad4f38e73682e5d6f46e56bd12

File: node/src/components/consensus/protocols/highway/round_success_meter.rs (hunk 5 above)
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/round_success_meter.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/8e16624e1fb6eac81369e4ed6679272be7fe2ac1

File: node/src/components/consensus/protocols/highway/round_success_meter.rs (hunk 6 above)
Comment (xcthulhu) [reply node/src/components/consensus/protocols/highway/round_success_meter.rs]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/8e16624e1fb6eac81369e4ed6679272be7fe2ac1

File: resources/local/chainspec.toml (hunk 4 above)
Comment (xcthulhu) [reply resources/local/chainspec.toml]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/88f1c9e89dd61c6f169a305d466f0b84e439d147

File: node/src/components/consensus/highway_core/state/tests.rs (hunk 1 above)
Comment (xcthulhu) [reply node/src/components/consensus/highway_core/state/tests.rs:177]: Fixed in https://github.com/CasperLabs/casper-node/pull/479/commits/0fca00cb55e7da746b6ffb954b020fd8d1ee9e22

**Pull request merged**